src/main/java/com/conveyal/r5/analyst/GridComputer.java

```diff
@@ -106,41 +119,107 @@ public void run() throws IOException {
-                    if (reqx < 0 || reqx >= request.width) continue; // outside of project bounds
-
-                    int index = reqy * request.width + reqx;
-                    if (times[index] < request.cutoffMinutes * 60) {
-                        access += grid.grid[gridx][gridy];
-                    }
-                }
+        PerTargetPropagater propagater =
+                new PerTargetPropagater(timesAtStopsEachIteration, nonTransferTravelTimesToStops, linkedTargets, request.request, request.cutoffMinutes * 60);
+
+        // compute the percentiles
+        double[] samples = new double[BOOTSTRAP_ITERATIONS + 1];
+
+        propagater.propagate((target, reachable) -> {
+            int gridx = target % grid.width;
+            int gridy = target / grid.width;
+            double opportunityCountAtTarget = grid.grid[gridx][gridy];
+
+            if (opportunityCountAtTarget < 1e-6) return; // don't bother with destinations that contain no opportunities
 
+            // index the reachable iterations, so we can skip over the non-reachable ones in bootstrap computations
+            // this improves computation speed (verified)
+            TIntList reachableInIterationsList = new TIntArrayList();
+
+            for (int i = 0; i < reachable.length; i++) {
+                if (reachable[i]) reachableInIterationsList.add(i);
             }
-            return Math.round(access);
+
+            int[] reachableInIterations = reachableInIterationsList.toArray();
+
+            boolean foundUnreachable = reachableInIterations.length < timesAtStopsEachIteration.length;
+            boolean foundReachable = reachableInIterations.length > 0;
+
+            if (foundReachable && foundUnreachable) {
+                // This origin is sometimes reachable within the time window, do bootstrapping to determine
+                // the distribution of how often
+                BOOTSTRAP:
+                for (int bootstrap = 0; bootstrap < BOOTSTRAP_ITERATIONS + 1; bootstrap++) {
+                    int count = 0;
+                    for (int iteration : reachableInIterations) {
+                        count += bootstrapWeights[bootstrap][iteration];
+                    }
+
+                    if (count > timesAtStopsEachIteration.length / 2) {
+                        samples[bootstrap] += opportunityCountAtTarget;
+                    }
+                }
+            } else if (foundReachable && !foundUnreachable) {
+                // this destination is always reachable and will be included in all bootstrap samples, no need to do the
+                // bootstrapping
+                for (int i = 0; i < samples.length; i++) samples[i] += grid.grid[gridx][gridy];
+            } // otherwise, this destination is never reachable, no need to do bootstrapping or increment samples
         });
 
+        int[] intSamples = DoubleStream.of(samples).mapToInt(d -> (int) Math.round(d)).toArray();
```

Comment:
I think I'm going to leave this and add a comment because this way it's explicit that we're rounding rather than casting/flooring.